File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        def dfs(i, j):
            if (i < 0 or j < 0 or i >= n or j >= m  
                or grid[i][j] == 0):
                return 0
            val = grid[i][j]
            grid[i][j] = 0
            maxi = 0
            maxi = max(
                dfs(i+1, j),
                dfs(i-1, j),
                dfs(i, j-1),
                dfs(i, j+1)
            )
            grid[i][j] = val
            return val + maxi

        
        maxi = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] != 0:
                    maxi = max(maxi, dfs(i, j))

        return maxi


        

File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.java
class Solution {
    private static final int [][] DIRS = {{0,-1}, {-1,0}, {1,0}, {0,1}};

    int dfs(int i , int j , int [][]g, int n, int m, Set<String> vis) {

        int val = g[i][j];
        g[i][j] = 0;
        int maxGoldFromHere = 0;

        for (int []dir: DIRS) {
            int r = i + dir[0];
            int c = j + dir[1];
            if (r >= 0 && c >= 0 && c < m && r < n && g[r][c] != 0) {
                maxGoldFromHere = Math.max(maxGoldFromHere, dfs(r, c, g, n, m, vis));
            } 
        }
        g[i][j] = val;
        return val + maxGoldFromHere;

    }

    public int getMaximumGold(int[][] g) {

        int n = g.length, m = g[0].length;
        int res = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (g[i][j] != 0) {
                    Set<String> vis = new HashSet<>();
                    res = Math.max(res, dfs(i, j, g, n, m, vis));
                }
            }
        }

        return res;
    }
}